libnet/ds, libnet/config: various cleanups

akerouanton opened the pull request:
**- What I did**

**libnet/i/kv/boltdb: remove unused field 'timeout'**

**libnetwork: unit tests: drop OptionBoltdbWithRandomDBFile**

This option fn was defining a custom directory, file name and bucket name for boltdb. Users can only change data-dir through `OptionDataDir`. Better reuse that function instead, that'll make refactorings easier.

It won't set a custom bucket name or file name as `OptionBoltdbWithRandomDBFile` was doing, but that's not needed since every test will use a different temp dir anyway.

**libnet/ds: simplify datastore.New()**

That function was needlessly complex. Instead of relying on a struct and a sub-struct, it now just takes two string params: a path and a bucket name.

Libnetwork config is now initialized with default values. A new struct is introduced in libnetwork/config to let tests customize the path and bucket name.

**libnet/i/kv/boltdb: fail fast in case of contention**

Make sure an error is returned straight away if there's contention on the underlying db file. This makes sure we don't reintroduce the issue fixed in d21d088, and it will help detect contention in parallelized tests if they're badly written. It effectively adds a new error mode to the daemon, but if anyone faces this error, they should fix their process manager.

**- How to verify it**

CI.

**- A picture of a cute animal (not mandatory but encouraged)**

akerouanton committed:
libnet/ds: simplify datastore.New()

That function was needlessly complex. Instead of relying on a struct and
a sub-struct, it now just takes two string params: a path and a bucket
name.

Libnetwork config is now initialized with default values. A new struct
is introduced in libnetwork/config to let tests customize the path and
bucket name.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet/i/kv/boltdb: remove unused field 'timeout'

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet/i/kv/boltdb: fail fast in case of contention

Make sure an error is returned straight away if there's contention on
the underlying db file. This makes sure we don't reintroduce the issue
fixed in d21d088, and it will help detect contention in parallelized
tests if they're badly written. It effectively adds a new error mode to
the daemon, but if anyone faces this error, they should fix their
process manager.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah commented on libnetwork/datastore/datastore.go:
```diff
@@ -109,15 +102,15 @@ func Key(key ...string) string {
 func New(cfg ScopeCfg) (*Store, error) {
-	if cfg.Client.Path == "" {
+	if cfg.Path == "" {
 		cfg = DefaultScope("")
```
Was looking at this code, and this is actually a weird construct; if we don't set a Path in the config, we IGNORE everything else, and just overwrite the whole `cfg`.

Wondering if the better signature would be to have `New(path string, cfg ScopeCfg)`

- `cfg` could be a pointer, so that `nil` can be passed to have the constructor use defaults
- export the default path constant

The latter one should be changed; currently we have a `defaultPrefix` const, which is only used in 1 location, and _always_ suffixed with `local-kv.db`; we could make that const `DefaultPath` (e.g.) and make passing a path a _required_ option 🤔 https://github.com/moby/moby/blob/98fecb0d54994b3b0c645b9b721651b03198bf9e/libnetwork/datastore/datastore.go#L77-L83

thaJeztah commented on libnetwork/datastore/datastore.go:
```diff
@@ -70,16 +70,17 @@ var (
 
-// DefaultScope returns a default scope config for clients to use.
-func DefaultScope(dataDir string) ScopeCfg {
+// NewConfig returns a new datastore Config, with the boltdb stored in dataDir.
+// If dataDir is empty, a default value is used.
+func NewConfig(dataDir string) Config {
 	var dbpath string
 	if dataDir == "" {
```
Hm... also see my comment on the previous commit; so if `dataDir` is the ONLY parameter that can be passed; do we need this function at all?

- Change `New()` to only accept a single argument (`path`)? 
- ☝️ Are we using the _dir_ anywhere, or only the full path to the _file_ ?
- `NewConfig` can be removed; it'd be part of the `New` constructor
- Maybe `Config` can even be un-exported if it's only used internally

thaJeztah commented on libnetwork/libnetwork_linux_test.go:
```diff
@@ -45,7 +45,7 @@ const (
 func TestMain(m *testing.M) {
 	// Cleanup local datastore file
-	_ = os.Remove(datastore.DefaultScope("").Path)
+	_ = os.Remove(datastore.NewConfig("").Path)
```
See my other comment; looks like if we export `DefaultPath`, then this code wouldn't have to call `NewConfig` and could just remove the `DefaultPath` (which would be `/var/lib/docker/network/files/local-kv.db`)

thaJeztah commented on libnetwork/config/config.go:
```diff
@@ -39,7 +39,7 @@ type Config struct {
 	NetworkControlPlaneMTU int
 	DefaultAddressPool     []*ipamutils.NetworkToSplit
-	Scope                  datastore.Config
+	DatastoreConfig        datastore.Config
```
And (with my other comments); this could become a string, and the location of the `lokal-kv.db`

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -20,6 +21,12 @@ var (
 
+// Config contains the options for a BoltDB bucket.
+type Config struct {
+	ConnectionTimeout time.Duration
```
Interesting; didn't an earlier commit remove an option to set the timeout? Should we have wired it up instead of removing? Looks like we hard-code it to 1 minute; is that something that should be configurable?

If we don't need it to be configurable, we could even remove this field

Even further; it looks like `Bucket` is also hardcoded to `libnetwork` 

❓ Do we need this Config struct _at all_ ? All options are hard-coded!

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -32,7 +39,7 @@ type BoltDB struct {
+func New(endpoint string, options *Config) (store.Store, error) {
 	if (options == nil) || (len(options.Bucket) == 0) {
 		return nil, ErrBoltBucketOptionMissing
 	}
```
And in that case; this would never return an error 😄

thaJeztah commented on libnetwork/store_test.go:
```diff
@@ -76,6 +76,6 @@ func OptionBoltdbWithRandomDBFile(t *testing.T) config.Option {
 	return func(c *config.Config) {
 		c.DatastoreConfig.Path = tmp
-		c.DatastoreConfig.Config = store.Config{Bucket: "testBackend"}
+		c.DatastoreConfig.Config = boltdb.Config{Bucket: "testBackend"}
```
Hm... ok; perhaps we'd still need some way to set it, but it look like _ONLY_ in tests? And _ONLY_ internal to the package, so .. perhaps we can wire that up in tests only.

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -21,12 +21,6 @@ var (
-	ConnectionTimeout time.Duration
-	Bucket            string
-}
-
```
Ah! And it's gone 😂 🎉

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -15,8 +15,10 @@ import (
 var (
-	// ErrBoltBucketOptionMissing is thrown when boltBcuket config option is missing
-	ErrBoltBucketOptionMissing = errors.New("boltBucket config option missing")
+	// ErrBoltPathOptionMissing is returned when New is called with an empty path
+	ErrBoltPathOptionMissing = errors.New("path to the boltdb is missing")
+	// ErrBoltBucketOptionMissing is returned when the bucket name is empty
+	ErrBoltBucketOptionMissing = errors.New("boltdb's bucket name is missing")
```
We should remove these errors altogether; they're only used in a single place, and exporting them implies that they're sentinel errors, but they're not used for that.

We can inline the errors, and if needed (not sure if these errors bubble up) change them to be an `errdefs.System` or `errdefs.InvalidParameter`.

It they're not user-configurable, then they should be an `errdefs.System`

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -19,6 +19,8 @@ var (
 	ErrBoltPathOptionMissing = errors.New("path to the boltdb is missing")
 	// ErrBoltBucketOptionMissing is returned when the bucket name is empty
 	ErrBoltBucketOptionMissing = errors.New("boltdb's bucket name is missing")
+	// ErrDBAlreadyOpen signals that this boltdb is already open by another
+	ErrDBAlreadyOpen = errors.New("the underlying boltdb file is already held open")
```
See my previous comment; if this is not used as a sentinel error, we shouldn't export it, and just inline the error.

thaJeztah commented on libnetwork/datastore/datastore.go:
```diff
@@ -102,7 +99,7 @@ func Key(key ...string) string {
 
 // New creates a new Store instance.
 func New(cfg Config) (*Store, error) {
-	s, err := boltdb.New(cfg.Path, cfg.Bucket, cfg.ConnectionTimeout)
```
This was just added in the previous commit, correct? Perhaps amend that commit 🤔

thaJeztah reviewed (commented):
Nice!! I left a couple of comments; 

- ⚠️ I went through commit-by-commit, so best to look through them as such, but
- 👉 some of my comments may have become obsolete in commits after that.
- 💡 it may still make sense to look at my "outdated" comments, in case they provide alternative options.

akerouanton committed:
libnetwork: unit tests: drop OptionBoltdbWithRandomDBFile

This option fn was defining a custom directory, file name and bucket
name for boltdb. Users can only change data-dir through `OptionDataDir`.
Better reuse that function instead, that'll make refactorings easier.

It won't set a custom bucket name or file name as `OptionBoltdbWithRandomDBFile`
was doing, but that's not needed since every test will use a different
temp dir anyway.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah commented on libnetwork/store_test.go:
```diff
@@ -5,17 +5,15 @@ import (
+		func(c *config.Config) { c.DatastoreBucket = bucket },
+	}
 
 	cfgOptions = append(cfgOptions, config.OptionDriverConfig("host", map[string]interface{}{
```
Minor nit; it looks like `cfgOptions` is never conditionally modified, so we only use it as an intermediate slice. So technically this could be;


```go
testController, err := New(
    config.OptionDataDir(path),
    func(c *config.Config) { c.DatastoreBucket = bucket },
    config.OptionDriverConfig("host", map[string]interface{}{
        netlabel.GenericData: options.Generic{},
    }),
) 
````

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -34,9 +35,20 @@ func New(path, bucket string) (store.Store, error) {
 	})
 	if err != nil {
+		if errors.Is(err, bolt.ErrTimeout) {
+			return nil, errors.New("the underlying boltdb file is already held open")
```
Should we include the path as part of the error message? (does the original `err` contain useful information, or is that just "timeout" without more context?)

thaJeztah commented on libnetwork/internal/kvstore/boltdb/boltdb.go:
```diff
@@ -34,9 +36,20 @@ func New(path, bucket string) (store.Store, error) {
 
 	db, err := bolt.Open(path, filePerm, &bolt.Options{
-		Timeout: time.Minute,
+		// The bbolt package opens the underlying db file and then issues an
+		// exclusive flock to ensures that it can safely write to the db. If
+		// it fails, it'll re-issue flocks every few ms until Timeout is
+		// reached.
+		// This nanosecond timeout bypasses that retry loop and make sure the
+		// bbolt package returns an ErrTimeout straight away. That way, the
+		// daemon, and unit tests, will fail fast and loudly instead of
+		// silently introducing delays.
+		Timeout: time.Nanosecond,
 	})
 	if err != nil {
+		if errors.Is(err, bolt.ErrTimeout) {
+			return nil, fmt.Errorf("boltdb file %s is already open", path)
+		}
```
Could use extra eyes on this logic, just to be sure 🫶

thaJeztah reviewed (approved):
LGTM, thanks!
